### PR TITLE
Rename backendStoreDriver to dataEngine

### PR DESCRIPTION
**Pull request by derekbit**
longhorn/longhorn#7095
https://github.com/longhorn/longhorn/issues/7480

deps on https://github.com/longhorn/backupstore/pull/153

### CODE REVIEW

File: pkg/imrpc/common.proto [hunk 1]
```diff
@@ -8,3 +8,8 @@ enum BackendStoreDriver {
     v1 = 0;
     v2 = 1;
 }
+
+enum DataEngine {
+    engine_v1 = 0;
+    engine_v2 = 1;
+}
```
Comment (innobead) [inline pkg/imrpc/common.proto:15]: Why not? so you don't need to do the conversion later.

```protobuf
enum DataEngine {
  v1 = 0;
  v2 = 1;
}
```

File: pkg/api/instance.go [hunk 2]
```diff
@@ -7,6 +7,13 @@ import (
 	"github.com/longhorn/longhorn-spdk-engine/proto/spdkrpc"
 )
 
+var (
+	dataEngines = map[string]string{
+		"engine_v1": "v1",
+		"engine_v2": "v2",
+	}
+)
```
Comment (innobead) [inline pkg/api/instance.go:15]: Just use v1/v2 in proto, so don't need to do a conversion here.

**Review by innobead — CHANGES_REQUESTED** [review]

File: pkg/imrpc/common.proto (hunk 1 above)
Comment (derekbit) [reply pkg/imrpc/common.proto:15]: protobuf doesn't allow the duplicate names....

File: pkg/api/instance.go (hunk 2 above)
Comment (derekbit) [reply pkg/api/instance.go:15]: The conversion is due to https://github.com/longhorn/longhorn-instance-manager/pull/348#discussion_r1437964221

File: pkg/imrpc/common.proto (hunk 1 above)
Comment (derekbit) [reply pkg/imrpc/common.proto:15]: ref: https://github.com/protocolbuffers/protobuf/issues/5425

Comment (innobead) [reply pkg/imrpc/common.proto:15]: I see, then let's follow the convention.

https://protobuf.dev/programming-guides/style/#enums

```protobuf
enum DataEngine {
  DATA_ENGINE_V1 = 0;
  DATA_ENGINE_V2 = 1;
}
```

Comment (derekbit) [reply pkg/imrpc/common.proto:15]: Updated. Thank you.

File: pkg/proxy/backup.go [hunk 3]
```diff
@@ -210,26 +210,26 @@ func (p *Proxy) BackupRestore(ctx context.Context, req *rpc.EngineBackupRestoreR
 		"serviceURL":         req.ProxyEngineRequest.Address,
 		"engineName":         req.ProxyEngineRequest.EngineName,
 		"volumeName":         req.ProxyEngineRequest.VolumeName,
-		"backendStoreDriver": req.ProxyEngineRequest.BackendStoreDriver,
+		"backendStoreDriver": req.ProxyEngineRequest.DataEngine,
```
Comment (c3y1huang) [inline pkg/proxy/backup.go:213]:
```suggestion
		"dataEngine": req.ProxyEngineRequest.DataEngine,
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: pkg/client/proxy_replica.go [hunk 4]
```diff
@@ -72,19 +74,21 @@ func (c *ProxyClient) ReplicaList(backendStoreDriver, engineName, volumeName,
 		return nil, errors.Wrap(err, "failed to list replicas for volume")
 	}
 
-	driver, ok := rpc.BackendStoreDriver_value[backendStoreDriver]
+	driver, ok := rpc.DataEngine_value[getDataEngine(dataEngine)]
```
Comment (innobead) [inline pkg/client/proxy_replica.go:77]: nit: Suggest creating a function for decoding `DecodeDataEngineEnum` (add to `imrpc` pkg), so there won't be a need to do conversion in each method.

```go
	dataEngine, ok := imrpc.DecodeDataEngineEnum(dataEngine)
	if !ok {
		return nil, fmt.Errorf("failed to list replicas for volume: invalid data engine %v", dataEngine)
	}

	defer func() {
		err = errors.Wrapf(err, "%v failed to list replicas for volume", c.getProxyErrorPrefix(serviceAddress))
	}()

	req := &rpc.ProxyEngineRequest{
		Address:            serviceAddress,
		EngineName:         engineName,
		Address:    serviceAddress,
		EngineName: engineName,
		// Deprecated
		BackendStoreDriver: rpc.BackendStoreDriver(driver),
		DataEngine:         dataEngine,
		VolumeName:         volumeName,
	}
```

**Review by innobead — DISMISSED** [review]
LGTM.

File: pkg/imrpc/proxy.pb.go [hunk 5]
```diff
@@ -31,10 +31,12 @@ type ProxyEngineRequest struct {
...
 	unknownFields protoimpl.UnknownFields
 
-	Address            string             `protobuf:"bytes,1,opt,name=address,proto3" json:"address,omitempty"`
+	Address string `protobuf:"bytes,1,opt,name=address,proto3" json:"address,omitempty"`
+	// Deprecated: Do not use.
 	BackendStoreDriver BackendStoreDriver `protobuf:"varint,2,opt,name=backend_store_driver,json=backendStoreDriver,proto3,enum=imrpc.BackendStoreDriver" json:"backend_store_driver,omitempty"`
 	EngineName         string             `protobuf:"bytes,3,opt,name=engine_name,json=engineName,proto3" json:"engine_name,omitempty"`
 	VolumeName         string             `protobuf:"bytes,4,opt,name=volume_name,json=volumeName,proto3" json:"volume_name,omitempty"`
+	DataEngine         DataEngine         `protobuf:"varint,5,opt,name=data_engine,json=dataEngine,proto3,enum=imrpc.DataEngine" json:"data_engine,omitempty"`
 }
```
[the remark this answers is not in the document]
Comment (innobead) [reply pkg/imrpc/proxy.pb.go:40]: I believe the comment is automatically generated. For those generated codes, no need to change them.

[the remark this answers is not in the document]
Comment (c3y1huang) [reply pkg/imrpc/proxy.pb.go:40]: I see. Didn't notice this is a generated code. Thank you for pointing out.

File: pkg/imrpc/instance.pb.go [hunk 6]
```diff
@@ -401,11 +411,13 @@ type InstanceDeleteRequest struct {
...
 	unknownFields protoimpl.UnknownFields
 
+	// Deprecated: Do not use.
 	BackendStoreDriver BackendStoreDriver `protobuf:"varint,1,opt,name=backend_store_driver,json=backendStoreDriver,proto3,enum=imrpc.BackendStoreDriver" json:"backend_store_driver,omitempty"`
 	Name               string             `protobuf:"bytes,2,opt,name=name,proto3" json:"name,omitempty"`
 	Type               string             `protobuf:"bytes,3,opt,name=type,proto3" json:"type,omitempty"`
 	DiskUuid           string             `protobuf:"bytes,4,opt,name=disk_uuid,json=diskUuid,proto3" json:"disk_uuid,omitempty"`
 	CleanupRequired    bool               `protobuf:"varint,5,opt,name=cleanup_required,json=cleanupRequired,proto3" json:"cleanup_required,omitempty"`
+	DataEngine         DataEngine         `protobuf:"varint,6,opt,name=data_engine,json=dataEngine,proto3,enum=imrpc.DataEngine" json:"data_engine,omitempty"`
 }
```
[the remark this answers is not in the document]
Comment (c3y1huang) [reply pkg/imrpc/instance.pb.go:421]: Clarified in https://github.com/longhorn/longhorn-instance-manager/pull/348#discussion_r1438056924.

File: pkg/client/instance.go [hunk 7]
```diff
@@ -85,6 +85,7 @@ type ReplicaCreateRequest struct {
 
 type InstanceCreateRequest struct {
 	BackendStoreDriver string
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/client/instance.go]: Added. Thank you.

File: pkg/imrpc/proxy.pb.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/imrpc/proxy.pb.go]: Added. Thank you.

File: pkg/api/instance.go [hunk 8]
```diff
@@ -26,9 +34,11 @@ type Instance struct {
 
 func RPCToInstance(obj *rpc.InstanceResponse) *Instance {
 	instance := &Instance{
-		Name:               obj.Spec.Name,
-		Type:               obj.Spec.Type,
+		Name: obj.Spec.Name,
+		Type: obj.Spec.Type,
+		// Deprecated
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/api/instance.go:41]: Added. Thank you.

File: pkg/proxy/backup.go (hunk 3 above)
Comment (derekbit) [reply pkg/proxy/backup.go]: Fixed. Thank you.

File: pkg/client/proxy_replica.go (hunk 4 above)
Comment (derekbit) [reply pkg/client/proxy_replica.go:77]: Currently, we need to decode `BackendStoreDriver` and `DataEngine`, so the codes will be length. I will add decode function after removing `BackendStoreDriver` in v1.7.0.

**Review by innobead — APPROVED** [review]

**Pull request merged**